api/src/services/authorization.js

```diff
@@ -15,59 +15,90 @@ const couchDbUser = doc => doc.type === 'user-settings';
 const getUserSettingsId = username => `org.couchdb.user:${username}`;
 
 const getDepth = (userCtx) => {
+  const depth = {
+    contactDepth: -1,
+    reportDepth: -1,
+  };
+
   if (!userCtx.roles || !userCtx.roles.length) {
-    return -1;
+    return depth;
   }
 
   const settings = config.get('replication_depth');
   if (!settings) {
-    return -1;
+    return depth;
   }
-  let depth = -1;
+
   userCtx.roles.forEach(function(role) {
     // find the role with the deepest depth
     const setting = settings.find(setting => setting.role === role);
     const settingDepth = setting && parseInt(setting.depth, 10);
-    if (!isNaN(settingDepth) && settingDepth > depth) {
-      depth = settingDepth;
+    if (!isNaN(settingDepth) && settingDepth > depth.contactDepth) {
+      depth.contactDepth = settingDepth;
+
+      const settingsReportDepth = setting && parseInt(setting.report_depth);
+      depth.reportDepth = !isNaN(settingsReportDepth) ? settingsReportDepth : -1;
     }
   });
+
   return depth;
 };
 
+const usesReportDepth = (authorizationContext) => authorizationContext.reportDepth >= 0;
+
 const hasAccessToUnassignedDocs = (userCtx) => {
   return config.get('district_admins_access_unallocated_messages') &&
          auth.hasAllPermissions(userCtx, 'can_view_unallocated_data_records');
 };
 
-const include = (array, ...values) => {
-  let newValues = 0;
-  values.forEach(value => value && array.indexOf(value) === -1 && array.push(value) && newValues++);
-  return newValues;
+const includeSubjects = (authorizationContext, newSubjects, depth) => {
+  const initialSubjectsCount = authorizationContext.subjectIds.length;
+  newSubjects.forEach(subject => {
+    if (!subject) {
+      return;
+    }
+
+    authorizationContext.subjectsDepth[subject] = depth;
+    authorizationContext.subjectIds.push(subject);
+  });
+  authorizationContext.subjectIds = _.uniq(authorizationContext.subjectIds);
```

Comment:
I'm concerned about the performance impact of moving this `_.uniq` in to the loop on line 378. Instead can we move the `_uniq` and length check out to the invoking function, so it can be done outside of the loop?

Reply:
The reason I put this here is because I removed the existence check before pushing (which would have been iterating the array for each subject). Unsure which one is more performant :) 
I'll switch back to existence checking.